lib/Dialect/LTL/LTLFolds.cpp

```diff
@@ -0,0 +1,117 @@
+//===- LTLFolds.cpp -------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/LTL/LTLOps.h"
+#include "mlir/IR/PatternMatch.h"
+
+using namespace circt;
+using namespace ltl;
+using namespace mlir;
+
+/// Check if an attribute is an integer zero.
+static bool isConstantZero(Attribute attr) {
+  if (!attr)
+    return false;
+  if (auto intAttr = dyn_cast<IntegerAttr>(attr))
+    return intAttr.getValue().isZero();
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// DelayOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult DelayOp::fold(FoldAdaptor adaptor) {
+  // delay(s, 0, 0) -> s
+  if (adaptor.getDelay() == 0 && adaptor.getLength() == 0)
+    return getInput();
+
+  return {};
+}
+
+LogicalResult DelayOp::canonicalize(DelayOp op, PatternRewriter &rewriter) {
```

Comment:
Any reason to roll this and the next as code rather than patterns?

Reply:
None at all, besides the declarative rewrites being harder to setup. Bad excuse though :stuck_out_tongue:. Move these into tblgen patterns.